src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp

```diff
@@ -0,0 +1,265 @@
+    }
+
+    if (shouldBeAlive)
+    {
+        if (objectId == NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::CheckIfAlive: FAIL: the object should be alive.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::CheckIfAlive: object alive as expected ");
+            ClassID classId{0};
+            hr = pCorProfilerInfo->GetClassFromObject(objectId, &classId);
+            if (FAILED(hr))
+            {
+                _failures++;
+                printf("(FAIL: impossible to get class from object).\n");
+            }
+            else
+            {
+                String typeName = GetClassIDName(classId);
+                wcout << typeName.ToWString() << std::endl;
+            }
+        }
+    }
+    else
+    {
+        if (objectId != NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::CheckIfAlive: FAIL: the object should not be alive anymore.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::CheckIfAlive: object not alive as expected.\n");
+        }
+    }
+}
+
+HRESULT HandlesProfiler::GarbageCollectionFinished()
+{
+    SHUTDOWNGUARD();
+
+    if (!_isInduced)
+    {
+        return S_OK;
+    }
+
+    HRESULT hr = S_OK;
+    if (_gcCount == 1)
+    {
+        // weak should not be here anymore
+        CheckIfAlive(_weakHandle, false);
+
+        // the others should still be alive
+        CheckIfAlive(_strongHandle, true);
+        CheckIfAlive(_pinnedHandle, true);
+    }
+    else
+    if (_gcCount == 2)
+    {
+        if (_strongHandle == NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::GarbageCollectionFinished: FAIL: null strong handle.\n");
+            return S_OK;
+        }
+        if (_pinnedHandle == NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::GarbageCollectionFinished: FAIL: null pinned handle.\n");
+            return S_OK;
+        }
+
+        // Destroy strong and pinned handles so next GC will release the objects
+        HRESULT hr = pCorProfilerInfo->DestroyHandle(_strongHandle);
+        if (FAILED(hr))
+        {
+            _failures++;
+            printf("HandlesProfiler::GarbageCollectionFinished: FAIL: DestroyHandle failed for strong handle.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::GarbageCollectionFinished: strong handle destroyed.\n");
+        }
+
+        hr = pCorProfilerInfo->DestroyHandle(_pinnedHandle);
+        if (FAILED(hr))
+        {
+            _failures++;
+            printf("HandlesProfiler::GarbageCollectionFinished: FAIL: DestroyHandle failed for pinned handle.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::GarbageCollectionFinished: pinned handle destroyed.\n");
+        }
+    }
+    else
+    if (_gcCount == 3)
+    {
+        // TODO: Check that instances wrapped by strong and pinned handles are not here any more
```

Comment:
It is not supported to get objects from a destroyed handle 😄. When you destroy the handle it is returned to the pool of available handles and you would be potentially getting some random other object, or garbage.

If you wanted to verify that the object is indeed collected you could use a second weak handle and verify that it returns NULL. I don't think it's necessary to add that though.

Reply:
✅  Thanks for the details: I'm removing the comment and updating the test accordingly